### ISSUE TITLE
Multiple Express Servers does not send socket.io message

### ISSUE THREAD

**Issue by svarughese91** [socketio/socket.io-redis#286]
I have 2 express servers running socket.io and using redis to store the socketIDs.  I have a device1 connected to express server 1 and another device2 connected to express server 2.  When i try to send a message from device1 to device2 I notice that the socket.io server does pick up the correct socket id but when it tries to emit the message I get this error: TypeError: Cannot read property 'emit' of undefined

I am not sure how to send messages when using multiple nodes.  Here is my code below: 

    var socketIO = require('socket.io');
    var io = socketIO(server);
    var client;
    
    // Create a Redis client
    const redisPort = redisConfig.redisPort,
        redisHostname = redisConfig.redisHost,
    var redis = require('redis');

    var RedisStore = require('socket.io-redis');
    var adapter = RedisStore({  host: redisHostname, redisPort });
    io.adapter(adapter);
    client = adapter.pubClient;

    // When a client connects
    io.sockets.on('connection', function (socket) {
        // Log in a user with their provided username
        socket.on('login', function (data) {

 

     var timerCount = 0;
            var login = function () {
    client.get('user:' + data.rec, function (err, socketId) {
                            try {
                                io.sockets.connected[socketId].emit('message', data);
                            } catch (err) {
                                if (timerCount === 100) {
                                    console.log("Waiting for user to join chat timed out!");
                                }
                                // Cant find user - user has not joined yet
                                console.log(err);
                                timerCount++;
                                setTimeout(login, 2000);
                            }
                        });
}

    ├─┬ socket.io@2.0.4
    │ ├── socket.io-adapter@1.1.1
    │ ├─┬ socket.io-client@2.0.4
    │ │ ├── socket.io-parser@3.1.2 deduped
    │ └─┬ socket.io-parser@3.1.2
    ├─┬ socket.io-redis@5.2.0
    │ ├── socket.io-adapter@1.1.1 deduped

**Comment by olivier37520** [comment]
With your code you are trying to send data to a socket on same io server instead sending to cluster.
It's not a socket.io-redis issue.
You have this error (error: TypeError: Cannot read property 'emit' of undefined) because socketId is not a socket connected on the io server.
So try to change 
`io.sockets.connected[socketId].emit('message', data); // sending on local io server`
by
`io.to(socketId).emit('message', data); // sending through cluster`

**Comment by svarughese91** [comment]
ok thanks for that and i made that change.  Now I am not getting that error but the messages are not getting received by the device now

**Comment by svarughese91** [comment]
Do the express nodes have to be clustered?

**Comment by svarughese91** [comment]
@olivier37520 is there something that I am missing?

**Comment by olivier37520** [comment]
Of course you connect the same redis server.

**Comment by svarughese91** [comment]
@olivier37520  Yeah I am connected to the same redis server.  When it first starts there is no problem and it communicates fine.  Then when i try to restart the socket.io server the communication stops working.  I am not sure why, its connecting to the same socket.io server and its emiting to the correct socketID but the client doesnt seem to be getting the event. Any idea why even sending to the correct socketID the client does not pick up on the event?

**Comment by olivier37520** [comment]
@svarughese91 first can you compare the socketId retrieved by redis and the socket.id on the second server . if they are identical and code above still don't work, you can try a workaround like this:
just after connection event 
`socket.join('room_'+socket.id);`
and in the emit part
`io.to('room_'+socketId).emit('message', data);`

As this you have a room by socket.
